src/web/source.rs

```diff
@@ -193,10 +193,47 @@ pub fn source_browser_handler(req: &mut Request) -> IronResult<Response> {
     let storage = extension!(req, Storage);
     let config = extension!(req, Config);
 
+    // TODO should we use match_version here?
+    let archive_storage: bool = {
+        let rows = ctry!(
+            req,
+            conn.query(
+                "
+                SELECT archive_storage
+                FROM releases 
+                INNER JOIN crates ON releases.crate_id = crates.id
+                WHERE 
+                    name = $1 AND 
+                    version = $2
+                ",
+                &[&name, &version]
+            )
+        );
+
+        let row = rows.get(0).ok_or(Nope::VersionNotFound)?;
+
+        row.get::<_, bool>(0)
+    };
+
     // try to get actual file first
     // skip if request is a directory
     let file = if !file_path.ends_with('/') {
-        DbFile::from_path(&storage, &file_path, &config).ok()
+        if archive_storage {
+            DbFile::from_archive_path(
+                &storage, // TODO: unify finding archive names somewhere
+                &format!("sources/{0}/sources-{0}-{1}.zip", name, version),
```

Comment:
You could pass the name and version into `from_archive_path`, and only do string formatting in there.